### ISSUE TITLE
Font smoothing glitches with Nerd Fonts

### ISSUE THREAD

**Issue by kgraefe** [mintty/mintty#943]
When using Nerd Fonts together with full font smoothing I get small lines between the powerline characters. That is not the case when using partial smotthing, though.

The screenshots are taken on mintty 3.0.6 with [Meslo Nerd Font patched for Powerlevel10k](https://github.com/romkatv/powerlevel10k#recommended-meslo-nerd-font-patched-for-powerlevel10k), but I can reproduce it with a lot of fonts available on https://www.nerdfonts.com/ as well.

![screenshot_full_font_smoothing](https://user-images.githubusercontent.com/3797324/68948492-4d458500-07b8-11ea-9691-d9cc274a4ded.png) 
![settings_full_font_smoothing](https://user-images.githubusercontent.com/3797324/68948501-520a3900-07b8-11ea-9149-beb6694386d2.png)
![screenshot_partial_font_smoothing](https://user-images.githubusercontent.com/3797324/68948508-546c9300-07b8-11ea-9918-5b858ca2fe43.png)
![settings_partial_font_smoothing](https://user-images.githubusercontent.com/3797324/68948514-56365680-07b8-11ea-98fb-b630c08252f5.png)

**Comment by kgraefe** [comment]
My `.minttyrc`:
```
BoldAsFont=no
ForegroundColour=131,148,150
BackgroundColour=0,43,54
CursorColour=220,50,47
CursorType=block
Locale=de_DE
Charset=UTF-8
RightClickAction=paste
FontHeight=10
Term=xterm-256color
CursorBlinks=no
Transparency=off
BoldAsColour=yes
Black=7,54,66
BoldBlack=0,43,54
Red=220,50,47
BoldRed=203,75,22
Green=133,153,0
BoldGreen=88,110,117
Yellow=181,137,0
BoldYellow=101,123,131
Blue=38,139,210
BoldBlue=131,148,150
Magenta=211,54,130
BoldMagenta=108,113,196
Cyan=42,161,152
BoldCyan=147,161,161
White=238,232,213
BoldWhite=253,246,227
FontSmoothing=full
BellType=0
Language=de
Scrollbar=none
ClicksTargetApp=no
#RowSpacing=-2
ThemeFile=MySolarized
CopyAsRTF=no
AllowSetSelection=yes
Font=MesloLGS NF
FontWeight=400
FontRender=textout
```

**Comment by mintty** [comment]
Cannot reproduce, assuming the dotted lines are the affected characters and are meant to be spaces.
But maybe they're something else; please copy/paste and filter them through `od -t x1`.

Any reason for `FontRender=textout`?

**Comment by kgraefe** [comment]
Sorry for being unclear. The dotted lines are part of the theme. The glitch is between on the long edge of the triangles: it should be one area with background of the character before, without the green vertical line.
![detail](https://user-images.githubusercontent.com/3797324/68950613-cba42600-07bc-11ea-9d0f-8509479d2dae.png)

Regarding `FontRender=textout`: I tried a lot of fonts in the past few days and ran into #754 at one point. Removing this line makes no difference for my issue.

**Comment by mintty** [comment]
OK, a screenshot does not really help to reproduce the issue. Please find out how to `echo` or `cat` (e.g. a screen *log* file) to demonstrate the issue.

**Comment by kgraefe** [comment]
Please find attached the log file (you can `cat` it).

I also managed to separete the actual texts:
```
> echo -e '\e[1m\e[38;5;254m\e[44m\e[38;5;255mtmp\e[0m\e[38;5;255m\e[44m\e[44m\e[38;5;254m\e[0m\e[38;5;255m\e[44m\e[44m\e[38;5;254m\e[0m\e[38;5;254m\e[44m\e[44m\e[38;5;254m \e[0m\e[38;5;254m\e[44m\e[43m\e[34m\xEE\x82\xB0\x0D\x0A\e[0'

> echo -e '\e[43m\e[30m?7\e[0m\e[0m\e[30m\e[43m\e[43m\e[30m \e[0m\e[30m\e[43m\e[49m\e[33m\xEE\x82\xB0\x0D\x0A\e[0m'
```


[zsh.log](https://github.com/mintty/mintty/files/3852132/zsh.log)

**Comment by mintty** [comment]
Since the effect varies with font smoothing, I'd say it's a font smoothing effect :)
If I double the triangle character, or repeat it with a black space between, the same "smoothed" bar effect occurs on every instance of the triangle. I'm not sure whether it's perhaps also a font issue; the glyph, as intended to reach out to the left cell border, might be extended by half a pixel to the left, possibly overriding the effect. In any case I don't see how the terminal could change the rendering, which is done by Windows in the end.

**Comment by kgraefe** [comment]
I'm not sure if that is helpful, but [ConEmu](https://github.com/Maximus5/ConEmu) does not have this issue.

![screenshot_conemu](https://user-images.githubusercontent.com/3797324/69058187-7c0b6780-0a13-11ea-9535-de2298a2c250.png)
![settings_conemu](https://user-images.githubusercontent.com/3797324/69058191-7dd52b00-0a13-11ea-8131-5af9d97c6a64.png)

**Comment by mintty** [comment]
If you zoom up the screenshots, you see that mintty applies font smoothing to the triangle symbol, while conemu apparently does not.
![943a](https://user-images.githubusercontent.com/12740416/69060701-15d51380-0a18-11ea-9ab6-f52e7f00e6b6.png)

Maybe it has some special rule to exclude certain symbols from font smoothing?
If you can find that out, we might come up with a strategy.

**Comment by kgraefe** [comment]
ConEmu supports those glyphs even when a font is selected that does not have them. So my guess would be that they draw them themselves instead of rendering the font. I could not find that in the code, though.

While researching I found a [blog post](https://blog.grenzdebiel.dynv6.net/?p=504) showing that the problem exists in PuTTY too.

However, I found a [work-around](https://github.com/kgraefe/nerd-fonts/commit/c9ef46d6f823adf7996a1528630170b179b551e0) by moving the affected glyphs slightly to the left when generating the font using nerd-font's font-patcher. I cannot fully estimate the impact of that change, but it's good enough for now.

As for mintty I guess there is nothing you can do, so imho the issue can closed.

**Comment by mintty** [comment]
Well, in theory there are two things that could be done.
First, affected glyphs could be output without font smoothing. However, as font smoothing is a parameter to font creation in Windows, that would need to double the number of font resources and complicate their handling.
Second, mintty could of course also draw affected symbols itself. Attribute resources to manage such self-drawing (which is done for other symbols already) are almost exhausted, though, to be checked.

To check which symbols are affected, I guess those are the geometric ones shown in https://github.com/ryanoasis/powerline-extra-symbols#glyphs, meaning the range U+E0B0..U+E0BF, without U+E0B5 and U+E0B7 which makes it a bit quirky.

**Comment by mintty** [comment]
Also, I've added a pseudo-script name "Private" so you can configure "nerd fonts" as a fallback only for Private Use character ranges with option `FontChoice`.

**Comment by mintty** [comment]
Released 3.1.0.

**Comment by AdySan** [comment]
@kgraefe Sorry this question is slightly off-topic. How do you get your shell to look like that in the first place. I'm trying to get something like oh-my-bash working on windows/cmder

**Comment by kgraefe** [comment]
I am using the Zsh with https://github.com/romkatv/powerlevel10k.

**Comment by mintty** [comment]
Released 3.7.5 with some related tweaks.
